ckan/lib/webassets_tools.py

```diff
@@ -146,19 +148,37 @@ def include_asset(name: str) -> None:
             type_ = "script"
             break
     else:
-        log.warning("Undefined asset type: %s", urls)
+        log.warning("Undefined asset type: %s", urls_info)
         return
+    g._webassets[type_].extend(urls_info)
 
-    g._webassets[type_].extend(urls)
 
-
-def _to_tag(url: str, type_: AssetType) -> str:
+def _to_tag(asset_info: Tuple[str, dict[str, str]], type_: AssetType) -> str:
     """Turn asset URL into corresponding HTML tag.
     """
-    if type_ == "style":
-        return f'<link href="{url}" rel="stylesheet"/>'
+
+    url, attrs = asset_info
+
+    is_preload = attrs.get("rel") and attrs["rel"] == "preload" and attrs.get("as")
+
+    if type_ == "style" and "rel" not in attrs:
+        attrs["rel"] = "stylesheet"
+    elif type_ == "script" and "type" not in attrs and not is_preload:
+        attrs["type"] = "text/javascript"
+
+    attr_items = []
+    for key, value in attrs.items():
+        if not value:
+            attr_items.append(key)
+        else:
+            attr_items.append(f'{key}="{value}"')
```

Comment:
The key/value can be anything so we should escape them to prevent generating invalid html.